### PR TITLE
Deser using serde

### PR DESCRIPTION
**Pull request by ohadravid**
Adds (basic, and not at all complete) support for two kinds of `desr` operations:

```rust
fn it_desr_into_struct() {
    let com_con = COMLibrary::new().unwrap();
    let wmi_con = WMIConnection::new(com_con.into()).unwrap();

    let p_svc = wmi_con.svc();

    #[derive(Deserialize, Debug)]
    struct Win32_OperatingSystem {
        Caption: String,
        Name: String,
        CurrentTimeZone: i16,
        Debug: bool,
        EncryptionLevel: u32,
        ForegroundApplicationBoost: u8,
        LastBootUpTime: WMIDateTime,
    }

    let enumerator = wmi_con
        .query("SELECT * FROM Win32_OperatingSystem")
        .unwrap();

    for res in enumerator {
        let w = res.unwrap();
        let w: Win32_OperatingSystem = from_wbem_class_obj(&w).unwrap();
    }
}
```

and

```rust
fn it_desr_into_map() {
    let com_con = COMLibrary::new().unwrap();
    let wmi_con = WMIConnection::new(com_con.into()).unwrap();
    let p_svc = wmi_con.svc();

    let enumerator = wmi_con
        .query("SELECT * FROM Win32_OperatingSystem")
        .unwrap();

    for res in enumerator {
        let w = res.unwrap();

        let w: HashMap<String, Variant> = from_wbem_class_obj(&w).unwrap();
    }
}
```

There are a bunch of to-dos, and not all types are implemented, and not all the memory is cleared correctly, but it generally works.

Also the code needs to be.. organized somehow.

### CODE REVIEW

File: src/connection.rs [hunk 1]
```diff
@@ -129,25 +128,22 @@ impl WMIConnection {
...
-                    ptr::null_mut(),
-                    ptr::null_mut(),
-                    ptr::null_mut(),
-                    0,
-                    ptr::null_mut(),
-                    ptr::null_mut(),
-                    &mut p_svc,
-                )
-            )?;
+            check_hres((*self.loc()).ConnectServer(
```
Comment (omerbenamram) [inline src/connection.rs]: use consistent rustfmt configuration (maybe add a git-hook or something).
this is polluting the PR :(

File: src/datetime.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            formatter,
+            "a formatted date and time string or a unix timestamp"
+        )
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+    {
+        value.parse().map_err(|err| E::custom(format!("{}", err)))
```
Comment (omerbenamram) [inline src/datetime.rs]: does `visit_str` has to be generic over all possible implementors of `de::Error`?
Can we throw a nicer error here?

File: src/datetime.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: de::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(DateTimeVisitor)
+    }
+}
+
+#[allow(non_camel_case_types)]
+mod tests {
```
Comment (omerbenamram) [inline src/datetime.rs]: add a test for a malformed date case

File: src/error.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+impl Debug for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Debug::fmt(&self.err, f)
+    }
+}
+
+impl de::Error for Error {
+    #[cold]
```
Comment (omerbenamram) [inline src/error.rs]: ❄️

File: src/error.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+use std::io;
+use std::result;
+
+use serde::de;
+use serde::ser;
+
+use failure::format_err;
+use failure::Fail;
+
+pub struct Error {
```
Comment (omerbenamram) [inline src/error.rs]: I'm not sure what this module is for.
It doesn't seem to be useful for anything.

Maybe we can just stick to using std::Error and derive our enum around it?
like https://github.com/arcnmx/serde-value/blob/master/src/de.rs#L81

File: src/query.rs [hunk 6]
```diff
@@ -1,27 +1,38 @@
...
 use std::ptr;
 use std::ptr::Unique;
-use winapi::um::wbemcli::{IWbemLocator, IWbemServices, IWbemClassObject, CLSID_WbemLocator, IID_IWbemLocator, IEnumWbemClassObject};
+use std::slice;
+use widestring::WideCStr;
+use widestring::WideCString;
+use winapi::shared::minwindef::UINT;
+use winapi::shared::ntdef::LONG;
+use winapi::shared::ntdef::NULL;
+use winapi::shared::rpcdce::RPC_C_AUTHN_LEVEL_CALL;
```
Comment (omerbenamram) [inline src/query.rs]: consider using clion auto-nest imports to de-clutter some lines

File: src/query.rs [hunk 7]
```diff
@@ -67,8 +77,49 @@ impl<'a> Drop for QueryResultEnumerator<'a> {
     }
 }
 
+pub struct IWbemClassWrapper {
+    pub inner: Option<Unique<IWbemClassObject>>,
+}
+
+impl IWbemClassWrapper {
```
Comment (omerbenamram) [inline src/query.rs]: Add docs for this struct

File: src/variant.rs [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    UI8(u64),
+
+    Array(Vec<Variant>),
+}
+
+impl Variant {
+    pub fn from_variant(vt: VARIANT) -> Result<Variant, Error> {
+        let variant_type: VARTYPE = unsafe { vt.n1.n2().vt };
+
+        if variant_type as u32 & VT_ARRAY == VT_ARRAY {
```
Comment (omerbenamram) [inline src/variant.rs]: why is this separated from the match statement below?

File: src/variant.rs [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+    Bool(bool),
+
+    UI1(u8),
+    UI8(u64),
+
+    Array(Vec<Variant>),
+}
+
+impl Variant {
```
Comment (omerbenamram) [inline src/variant.rs]: I like this! this is a pretty good abstraction over the unsafe native variant.

File: src/variant.rs [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+                "Converting from variant type {:#X} is not implemented yet",
+                variant_type
+            ),
+        };
+
+        Ok(variant_value)
+    }
+}
+
+struct SeqAccess {
```
Comment (omerbenamram) [inline src/variant.rs]: Why is this needed instead of a generic `impl Iterator for X`? just curios.

Comment (ohadravid) [reply src/variant.rs]: You would be breaking the orphan rules

Comment (omerbenamram) [reply src/variant.rs]: Gotcha.

File: src/error.rs (hunk 4 above)
Comment (omerbenamram) [reply src/error.rs]: image

File: src/de/variant_de.rs [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+    fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>, Self::Error>
+        where
+            T: de::DeserializeSeed<'de>,
+    {
+        if self.i >= self.data.len() {
+            return Ok(None);
+        }
+
+        let res: Variant = self.data.swap_remove(self.i);
```
Comment (ohadravid) [inline src/de/variant_de.rs]: @omerbenamram BTW this is a terrible bug (since `swap_remove`, as the name suggest, does not preserve the order, so we basically just return some random half-list).

It works because the array used in the tests is of length 1 (🙄 ).

I fixed it in a newer branch.

Comment (omerbenamram) [reply src/de/variant_de.rs]: מצחיק
